sphinx/environment/__init__.py

```diff
@@ -668,6 +669,10 @@ def read_doc(self, docname, app=None):
         self.settings['trim_footnote_reference_space'] = \
             self.config.trim_footnote_reference_space
         self.settings['gettext_compact'] = self.config.gettext_compact
+        language = (self.config.language or 'en').replace('_', '-')
+        self.settings['language_code'] = language
+        if language in smartchars.quotes:
+            self.settings['smart_quotes'] = self.config.html_use_smartypants
```

Comment:
TL;DR: +1 for removing!

Certainly, LaTeX builder applies smartypants process to the document regardless. But it is not related with `html_use_smartypants`. The process is applied even if the config is disabled.
It is only used to switch the translator component to smarty-pants'ed one. So it only affects to HTML builder and its descendants (EPUB and so on).
```
# in sphinx/builders/html.py
    def init_translator_class(self):
        # type: () -> None
        if self.translator_class is None:
            if self.config.html_experimental_html5_writer and html5_ready:
                if self.config.html_use_smartypants:
                    self.translator_class = SmartyPantsHTML5Translator
                else:
                    self.translator_class = HTML5Translator
            else:
                if self.config.html_use_smartypants:
                    self.translator_class = SmartyPantsHTMLTranslator
                else:
                    self.translator_class = HTMLTranslator
```

Hence, as I commented above, this changes the behavior of the config value.

But nobody would want to change the settings of smarty quotes by builder as you said.
So the target of smartypants setting should be widen to whole of the builders.
And it is good option to integrate the settings to docutils'.
+1 for removing `html_use_smartypants` and moving to the `smarty_quotes` of docutils'.